src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -18,56 +19,64 @@ internal sealed class SnapshotManager
             /// </summary>
             private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
             /// <summary>
-            /// The dictionary key corresponds to Syntax position. The dictionary should be sorted in descending order.
+            /// The snapshot array should be sorted in ascending order by the position tuple element in order for the binary search algorithm to
+            /// function correctly.
             /// </summary>
-            private readonly ImmutableSortedDictionary<int, Snapshot> _incrementalSnapshots;
+            private readonly ImmutableArray<(int position, Snapshot snapshot)> _incrementalSnapshots;
```

Comment:
Consider upgrading the (int, snapshot) tuple into its own lightweight struct type, which helps to clarify the intent of the code. That could also be made comparable so you wouldn't need the explicit comparer.